application/app/Http/Controllers/Admin/BallotController.php

```diff
@@ -0,0 +1,239 @@
+<?php
+
+namespace App\Http\Controllers\Admin;
+
+use App\DataTransferObjects\BallotData;
+use App\DataTransferObjects\QuestionChoiceData;
+use App\DataTransferObjects\QuestionData;
+use App\Enums\ModelStatusEnum;
+use App\Enums\QuestionTypeEnum;
+use App\Http\Controllers\Controller;
+use App\Models\Ballot;
+use App\Models\BallotQuestionChoice;
+use App\Models\Question;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Redirect;
+use Inertia\Inertia;
+use Inertia\Response;
+use JetBrains\PhpStorm\NoReturn;
+use Momentum\Modal\Modal;
+use Illuminate\Support\Facades\Gate;
+
+class BallotController extends Controller
+{
+    /**
+     * Display the new ballot's form.
+     */
+    public function create(Request $request): Response
+    {
+        return Inertia::render('Auth/Ballot/Create', []);
+    }
+
+    /**
+     * Display the ballot's form.
+     */
+    public function view(Request $request, Ballot $ballot): Response
+    {
+        $ballot->load(['questions.choices']);
+        return Inertia::render('Auth/Ballot/View', [
+            'ballot' => BallotData::from($ballot),
+        ]);
+    }
+
+    /**
+     * Display the ballot's form.
+     */
+    public function edit(Request $request, Ballot $ballot): Response
+    {
+        $ballot->load(['questions.choices']);
+        return Inertia::render('Auth/Ballot/Edit', [
+            'ballot' => BallotData::from($ballot)
+        ]);
+    }
+
+    /**
+     * Store a newly created Ballot in storage.
+     */
+    public function store(BallotData $ballotData): RedirectResponse
+    {
+        $response = Gate::inspect('create', Ballot::class);
+
+        if ($response->allowed()) {
+            $ballot = new Ballot;
+            $ballot->fill($ballotData->all());
+            $ballot->save();
+
+            return Redirect::route('admin.ballots.view', ['ballot' => $ballot->hash]);
+        } else {
+            return Redirect::back()->withErrors(['error' => 'Not authorized to create ballot']);
+        }
+    }
+
+    /**
+     * Update the ballot's profile information.
+     */
+    public function update(Request $request, Ballot $ballot){
+
+        $response = $request->status == 'published' ? Gate::inspect('publish', $ballot) : Gate::inspect('update', $ballot);
+
+        if ($response->allowed()) {
+            $ballot = Ballot::byHash($ballot->hash);
+            $ballot->title = $request->title;
```

Comment:
I wondering if it would be a better approach to implement `FormRequest` and explicitly define acceptable and expected data along with the permission checking?

* https://laravel.com/docs/10.x/validation#form-request-validation
* https://laravel.com/docs/10.x/validation#authorizing-form-requests

Reply:
We are implementing FormRequest under the hood! 

`BallotData` is a magic [data](https://spatie.be/docs/laravel-data/v3/introduction) object that acts as a FormRequest object when injected in a controller. 

We should be injecting it here like we're doing on the [store](https://github.com/DripDropz/onchain-voting/pull/16/files/d00e701539e4f9fbea9fafaf75fafa2cdc7c267b#diff-5f2f4ca0a7cfa89672af7ef36907650038b667f9302def985e668e582f977646R59). 

I will update it!